wasmcloud-wasi-kv/wasmcloud-wasi-kv-wit/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+const ERROR_NO_SUCH_KEY: Error = 400;
+
+/// Unexpected Internal Error
+const ERROR_INTERNAL: Error = 500;
+/// Internal map space has been exhausted
+const ERROR_INTERNAL_MAP_SPACE_EXHAUSTED: Error = 501;
+/// An Error occurred while trying to access internal memory
+const ERROR_INTERNAL_MEMORY_ACCESS_ERROR: Error = 502;
+/// An error occurred wiht the upstream keyvalue store
+const ERROR_INTERNAL_UPSTREAM_OPERATION_FAILURE: Error = 503;
+const ERROR_INTERNAL_PROVIDER_REQUEST_SERIALIZATION_FAILURE: Error = 505;
+const ERROR_INTERNAL_PROVIDER_RESPONSE_DESERIALIZATION_FAILURE: Error = 506;
+
+/// This struct implements WASI interfaces
+struct KvProvider {}
+
+///////////////////////
+// Utility Functions //
+///////////////////////
+
+/// Helper function for cloning the value in a hashmap at a given key
+fn clone_hashmap_value<K: Hash + Eq, V: Clone>(
+    map: &'static RwLock<HashMap<K, V>>,
+    key: K,
+) -> Result<Option<V>, Error> {
+    let map_r = map.read()
+        .map_err(|_| ERROR_INTERNAL_MEMORY_ACCESS_ERROR)?;
+
+    let value = match map_r.get(&key) {
+        Some(v) => v,
+        None => {
+            return Err(ERROR_INTERNAL)
+        },
+    };
+
+    let mut map_w = map.write()
+        .map_err(|_| ERROR_INTERNAL_MEMORY_ACCESS_ERROR)?;
+    let inserted = map_w.insert(key, value.clone());
+
+    Ok(inserted)
+}
+
+/// Helper for incrementing an existing thread local
+fn get_next_available_key_u32<V>(
+    map: &'static RwLock<HashMap<u32, V>>,
+    start: &'static Mutex<u32>,
+) -> Result<u32, Error> {
+    let map = map.read().map_err(|_| ERROR_INTERNAL)?;
+    let mut start = start.lock().map_err(|_| ERROR_INTERNAL)?;
+
+    let mut idx = *start + 1;
+    while idx != *start {
+        match map.contains_key(&idx) {
+            true => {
+                idx = (idx % u32::MAX) + 1;
+            }
+            false => {
+                *start = idx;
+                return Ok(idx);
+            }
+        }
+    }
+    Err(ERROR_INTERNAL_MAP_SPACE_EXHAUSTED)
+}
+
+/// Create a new incoming value
+fn create_incoming_value(bytes: Vec<u8>) -> Result<IncomingValue, Error> {
+    let mut ivs = INCOMING_VALUES.write()
+        .map_err(|_| ERROR_INTERNAL_MEMORY_ACCESS_ERROR)?;
+    let next_iv = get_next_available_key_u32(&INCOMING_VALUES, &NEXT_INCOMING_VALUE)?;
+
+    let _ = ivs.insert(next_iv, bytes);
+
+    Ok(next_iv)
+}
+
+///////////////////////////////////////
+// KeyValue Interface Implementation //
+///////////////////////////////////////
+
+impl keyvalue::Keyvalue for KvProvider {
+    /// Get a value from the key value store
+    fn get(_bucket: Bucket, key: String) -> Result<IncomingValue, Error> {
+        println!("[debug][kv-provider] performing get...");
+
+        let mut encoder = wasmbus_rpc::cbor::vec_encoder(false);
+        encoder.str(&key)
+            .map_err(|_| ERROR_INTERNAL_PROVIDER_REQUEST_SERIALIZATION_FAILURE)?;
+        let payload = encoder.into_inner();
```

Comment:
Are you sure this is valid? I though everything that goes over wasmbus uses msgpack

Reply:
I'm not sure, but this is the only [encoder that is there](https://docs.rs/wasmbus-rpc/latest/wasmbus_rpc/common/enum.MessageFormat.html?search=encoder). I checked [the other codebases as well](https://github.com/wasmCloud/interfaces/blob/f6c80bb5212dd56dd99d69d579b5b952d894e3c3/testing/rust/src/testing.rs), and it looks to be the right thing.

Reply:
Hey @autodidaddict can you explain if this would work?

Also, [TIL CBOR was a fork of messagepack](https://github.com/msgpack/msgpack/issues/258)

Reply:
Sure, but why not just use `wasmbus::rpc::serialize` ?

Reply:
Ah you're right, will use serialize if it works there!

- [ ] use serialize instead of explicit encoding